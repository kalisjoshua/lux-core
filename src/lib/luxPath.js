/**
 * @module lib/luxPath
 * @memberof lux-lib
 */

import urlParse from './urlParse';

/**
 * @typedef {string} LuxPath
 * @global
 *
 * @description
 * Path string should always start with '/', but can include querystring
 * parameters; the pathname and querystring parameters will be passed along in
 * the API calls.
 */

/**
 * Consistent LuxPath factory - path(name) and querystring - should be
 * considered when retrieving the current LuxPath; since this is not included
 * in common APIs this function aims to provide the utility.
 *
 * @param  {string} loc - The URL to parse.
 *
 * @return {LuxPath} - The standardized luxPath value for the given URL.
 *
 * @example
 * import luxPath from './luxPath';
 *
 * const { pathname } = luxPath('http://foo.bar/baz/boo?fizz=buzz');
 * // pathname === '/baz/boo'
 */
function luxPath(loc) {
  const url = urlParse(`${loc}`);
  const path = url.pathname
    .replace(/\/+$/, ''); // remove trailing slashes

  return `${path}${url.search}` || '/';
}

export default luxPath;
