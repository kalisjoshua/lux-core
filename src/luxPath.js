/**
 * @module luxPath
 * @memberof lux
 */

import urlParse from 'url-parse';

import { isString } from './is';

/**
 * @typedef LuxPath
 * @type {String}
 * @global
 *
 * @description
 * Path string should always start with '/', but can include querystring
 * parameters; the pathname and querystring parameters will be passed along in
 * the API calls.
 */

/**
 * Consistent LuxPath factory - path(name) and querystring - should be
 * considered when retrieving the current LuxPath; since this is not included
 * in common APIs this function aims to provide the utility.
 *
 * @param  {String} loc - The URL to parse.
 *
 * @return {LuxPath} - The standardized luxPath value for the given URL.
 */
function luxPath(loc) {
  const url = isString(loc) ? urlParse(loc) : loc;
  const path = url.pathname.replace(/\/+$/, '');
  const query = url.query || '';

  return path + query;
}

export default luxPath;
