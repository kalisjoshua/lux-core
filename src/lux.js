/**
 * @module lux
 */

import apiRequest from './apiRequest';
import { isString } from './lib/is';
import luxPath from './luxPath';
import { responseModelFormat } from './responseModel';

let apiRoot;
const errorInvalidConfig = new Error('Lux must be configured before routing.');
const errorInvalidPath = new Error('Paths must be strings.');
const format = responseModelFormat;

function configRequired() {
  throw new Error('Lux requires an API root URI.');
}

function errorStr(prop, problem) {

  return `Configuration property \`${prop}\` ${problem}.`;
}

/**
 * Lux requires some configuration before routing will be possible; until valid
 * configuration is provided routing will not work.
 *
 * @param  {String} api - The API root URI.
 *
 * @return {Promise} - The first Promise handler will receive the ResponseModel
 * representation of the resource from the API; or an error.
 *
 * @example
 * import { init } from '@luxui/core';
 *
 * init({
 *   api: 'http://example.com',
 * });
 */
function init({ api } = configRequired()) {
  apiRoot = null;

  if (!isString(api)) {
    throw new Error(errorStr('api', 'is not a string'));
  }

  apiRoot = api;
}

/**
 * The `lux` function is for routing to - and retrieving API resource - pages.
 *
 * @param  {LuxPath} path - The route to load. The `path` will default to the
 * current `window.location` if none is provided.
 *
 * @return {Promise}
 *
 * @example
 * import lux from '@luxui/core';
 *
 * // after `init(config)`; start the application.
 * lux();
 *
 * @example
 * import lux from '@luxui/core';
 *
 * // route to the path given; if not already at that path
 * lux('/home');
 */
function lux(path) {
  let pending;

  if (!apiRoot) {
    pending = new Promise(resolve =>
      resolve(format({}, errorInvalidConfig)));
  } else if (!path || isString(path)) {
    pending = apiRequest(apiRoot + luxPath(path));
  } else {
    pending = new Promise(resolve =>
      resolve(format({}, errorInvalidPath)));
  }

  return pending;
}

export default lux;

export {
  init
};
