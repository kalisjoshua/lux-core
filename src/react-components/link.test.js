jest.mock('../index');
import lux from '../index';

jest.mock('../lib/luxPath');
import luxPath from '../lib/luxPath';

import React from 'react';
import renderer from 'react-test-renderer';

import Link, { clickHandler, isLeftClick, isModifiedClick } from './link';

describe('Lux - Link', function () {
  it('should match the snapshot', function () {
    const component = renderer.create(
      <Link href="/hello">Hello</Link>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('event handlers', function () {
    it('should have isLeftClick defined', function () {
      expect(typeof isLeftClick).toMatch(/function/i);
    });

    it('should have isModified defined', function () {
      expect(typeof isModifiedClick).toMatch(/function/i);
    });

    it('should detect left-click', function () {
      expect(isLeftClick({ button: 0 })).toBe(true);
    });

    it('should detect non left-click', function () {
      expect(isLeftClick()).toBe(false);
      expect(isLeftClick({})).toBe(false);
      expect(isLeftClick({ button: 1234 })).toBe(false);
    });

    [
      ['undefined event', undefined, false],
      ['empty event object', {}, false],
      ['alt key', { altKey: true }, true],
      ['ctrl key', { ctrlKey: true }, true],
      ['meta key', { metaKey: true }, true],
      ['shift key', { shiftKey: true }, true],
    ].forEach(set => {
      it(`should detect click modifiers (${set.shift()})`, function () {
        expect(isModifiedClick(set.shift())).toBe(set.shift());
      });
    });
  });

  describe('click handler', function () {
    it('should have clickHandler defined', function () {
      expect(typeof clickHandler).toMatch(/function/i);
    });

    it('should not navigate if the link is the same as the current page', function () {
      luxPath.mockReturnValue('/same');

      const event = {
        button: 0,
        preventDefault() {},
        target: {
          href: '/something-else',
          nodeName: 'A',
        }
      };

      clickHandler(event);
      expect(lux.mock.calls.length).toBe(0);
    });

    it('should navigate if the link is different than the current page', function () {
      luxPath
        .mockReturnValueOnce('/once')
        .mockReturnValueOnce('/twice');

      const event = {
        button: 0,
        preventDefault() {},
        target: {
          href: '/something-else',
          nodeName: 'A',
        }
      };

      spyOn(history, 'pushState');

      clickHandler(event);
      expect(lux.mock.calls.length).toBe(1);
      expect(history.pushState).toHaveBeenCalled();
    });

    it('should traverse up the DOM if Link contains children', function () {
      const link = '/nesting-link';

      luxPath
        .mockReturnValueOnce(link)
        .mockReturnValueOnce('/home');

      const event = {
        button: 0,
        preventDefault() {},
        target: {
          parentNode: {
            href: link,
            nodeName: 'A',
          }
        }
      };

      spyOn(history, 'pushState');

      clickHandler(event);
      expect(lux).lastCalledWith(link);
    });
  });
});
