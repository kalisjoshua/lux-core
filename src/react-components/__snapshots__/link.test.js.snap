exports[`Lux - Link should match the snapshot 1`] = `
<a
  href="/hello"
  onClick={[Function]}>
  Hello
</a>
`;
